src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java

```diff
@@ -0,0 +1,247 @@
+ * Design: Reproduces the signature verification algorithm used by Hart, which
+ * is the .NET Framework 4.8.1 implementation of a detached XML Digital Signature.
+ * Conditions: When verifying the signature of a Hart file.
+ * Version history: see https://github.com/BrightSpots/rcv.
+ */
+
+package network.brightspots.rcv;
+
+import static network.brightspots.rcv.SecurityXmlParsers.HartSignature;
+import static network.brightspots.rcv.SecurityXmlParsers.RsaKeyValue;
+import static network.brightspots.rcv.SecurityXmlParsers.SignedInfo;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.dataformat.xml.XmlMapper;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.RSAPublicKeySpec;
+import java.util.Base64;
+import javax.xml.crypto.OctetStreamData;
+import javax.xml.crypto.dsig.CanonicalizationMethod;
+import javax.xml.crypto.dsig.TransformException;
+import javax.xml.crypto.dsig.XMLSignatureFactory;
+import javax.xml.crypto.dsig.spec.C14NMethodParameterSpec;
+
+class SecuritySignatureValidation {
+  /**
+   * This function returns whether the signature matches validation.
+   * It throws an error if it is unable to perform this check, for reasons including:
+   * 1. The public key in the signed file does not match the expected key
+   * 2. The canonicalization or hashing algorithm is not supported
+   * 3. The corresponding file has been edited and its hash no longer
+   * matches what's found in signatureKeyFile
+   */
+  public static boolean verifyPublicKeySignature(
+          RsaKeyValue publicKey,
+          File signatureKeyFile,
+          File dataFile) throws CouldNotVerifySignatureException {
+    // Load the .sig.xml file into a HartSignature object
+    HartSignature hartSignature;
+    try {
+      hartSignature = readFromXml(signatureKeyFile, HartSignature.class);
+    } catch (IOException e) {
+      throw new CouldNotVerifySignatureException("Failed to read files: " + e.getMessage());
+    }
+
+    // Checks that will throw exceptions if they don't pass
+    ensurePublicKeyMatchesExpectedValue(hartSignature, publicKey, signatureKeyFile);
+    ensureDigestMatchesExpectedValue(dataFile, hartSignature);
+
+    // Check that might throw an exception if its unable to perform the signature validation.
+    return checkSignature(hartSignature, publicKey);
```

Comment:
Agree, the only downside is that we're allowed to throw from anywhere in a function, but we can't have multiple returns, so this would make these functions messier.

I guess I'll summarize the current logic as so:
1. We throw if we were unable to check the signature
2. We return false if the signature didn't match

I agree it's weird. I'm a little more partial to _always_ throwing, since that would make the code cleaner, and the requirements specify it must be a "halting exception," and when we return false we end up throwing anyway.

Reply:
I opted to throw two types of exceptions, rather than having one type of error return a boolean and another throw an exception. How's the updated code look?